src/mds/Server.cc

```diff
@@ -716,10 +716,24 @@ void Server::find_idle_sessions()
     Session *session = mds->sessionmap.get_oldest_session(Session::STATE_OPEN);
     if (!session) break;
     dout(20) << "laggiest active session is " << session->info.inst << dendl;
+
     if (session->last_cap_renew >= cutoff) {
       dout(20) << "laggiest active session is " << session->info.inst << " and sufficiently new (" 
 	       << session->last_cap_renew << ")" << dendl;
       break;
+    } else if (session->stale_mark_deferred and cutoff - session->
+		last_cap_renew < mds->mdsmap->get_session_autoclose()) {
+      break;
+    }
+
+    client_t client = session->get_client();
+    if (not (session->stale_mark_deferred or mds->locker->
+	is_revoking_any_caps_from(client) or session->
+	is_session_flushing_msgs(client.v))) {
+	dout(20) << "deferring marking client " << client.v << " stale since "
+		  "it holds no caps" << dendl;
+	session->stale_mark_deferred = true;
+	break;
```

Comment:
If the code hits this point, am I correct in thinking, that there might be a session after the current one that might have also crossed `session_timeout` limit and it might hold caps unlike this one? Such a client should be marked stale but it would be spared, I suppose. 

If I am correct, I should modify the loop to go on with second oldest client and so on until a session hit [this](https://github.com/rishabh-d-dave/ceph/blob/master/src/mds/Server.cc#L698) or [this](https://github.com/rishabh-d-dave/ceph/blob/master/src/mds/Server.cc#L703) line.

Reply:
yes. see my code https://github.com/ceph/ceph/pull/20538

Reply:
I went through the changes in `Server::find_idle_session` in #20538, I get it . To test locally, I am duplicating the changes introduced there in the `Server::find_idle_session`.  In that case, this PR would have wait for #20538 to merge. If there is a better to do this, let me know.